src/org/labkey/test/pages/query/SourceQueryPage.java

```diff
@@ -0,0 +1,95 @@
+package org.labkey.test.pages.query;
+
+import org.labkey.test.Locator;
+import org.labkey.test.WebDriverWrapper;
+import org.labkey.test.WebTestHelper;
+import org.labkey.test.pages.LabKeyPage;
+import org.labkey.test.util.DataRegion;
+import org.labkey.test.util.DataRegionTable;
+import org.labkey.test.util.Ext4Helper;
+import org.openqa.selenium.WebDriver;
+
+import java.util.Map;
+
+/**
+ * Page wrapper for 'QueryController.SourceQueryAction'
+ */
+public class SourceQueryPage extends LabKeyPage<SourceQueryPage.ElementCache>
+{
+    public SourceQueryPage(WebDriver driver)
+    {
+        super(driver);
+    }
+
+    public static SourceQueryPage beginAt(WebDriverWrapper webDriverWrapper, String schemaName, String queryName)
+    {
+        return beginAt(webDriverWrapper, webDriverWrapper.getCurrentContainerPath(), schemaName, queryName);
+    }
+
+    public static SourceQueryPage beginAt(WebDriverWrapper webDriverWrapper, String containerPath, String schemaName, String queryName)
+    {
+        webDriverWrapper.beginAt(WebTestHelper.buildURL("query", containerPath, "sourceQuery", Map.of("schemaName", schemaName, "query.queryName", queryName)));
+        return new SourceQueryPage(webDriverWrapper.getDriver());
+    }
+
+    public SourceQueryPage viewSource()
+    {
+        _ext4Helper.clickExt4Tab("Source");
+        return this;
+    }
+
+    public SourceQueryPage setSource(String sql)
+    {
+        viewSource();
+        setCodeEditorValue("queryText", "");
+        setCodeEditorValue("queryText", sql);
```

Comment:
I'm assuming that the double call here is to clear `queryText` and then set it to the value stored in `sql`. If that is not the case you may want to add a comment as to why the two calls.

Reply:
Was trying to figure out a problem with detecting when the data grid would refresh (this was not the solution); I've removed the set to empty